File: 9_Backtracking.cpp
#include <vector>
#include <iostream>
#include <queue>
#include <stack>
#include <unordered_map>
#include <map>
#include <string>
#include <algorithm>
#include <list>
#include <sstream>
#include <climits>
#include <unordered_set>

using namespace std;
/*
1. Subsets: https://leetcode.com/problems/subsets/description/
2. Subsets II: https://leetcode.com/problems/subsets-ii/description/
3. Combination Sum: https://leetcode.com/problems/combination-sum/description/
4. Combination Sum II: https://leetcode.com/problems/combination-sum-ii/description/
7. Word Search: https://leetcode.com/problems/word-search/description/
*/

// 1
class AllSubsets
{
    void subsets_helper(vector<int> &nums, int pos, vector<int> &subset, vector<vector<int>> &all_subsets)
    {
        // all the elements in nums have been processed
        if (pos == nums.size())
        {
            all_subsets.push_back(subset);
            return;
        }
        // don't take nums[pos] element
        subsets_helper(nums, pos + 1, subset, all_subsets);

        // take nums[pos] element
        subset.push_back(nums[pos]);
        subsets_helper(nums, pos + 1, subset, all_subsets);
        subset.pop_back(); // don't forget to remove it
    }

public:
    vector<vector<int>> subsets_iter(vector<int> &nums)
    {
        vector<vector<int>> ans;
        ans.push_back({});
        for (auto v : nums)
        {
            vector<vector<int>> temp;
            for (auto subset : ans)
            {
                temp.push_back(subset);
                subset.push_back(v);
                temp.push_back(subset);
            }
            ans = temp;
        }
        return ans;
    }
    vector<vector<int>> subsets(vector<int> &nums)
    {
        if (nums.size() == 0)
            return {};

        vector<vector<int>> all_subsets = {};
        vector<int> subset = {};
        subsets_helper(nums, 0, subset, all_subsets);
        return all_subsets;
    }
};

// 2
class AllUniqueSubsetsWithDuplicates
{
public:
    vector<vector<int>> subsetsWithDup(vector<int> &nums)
    {
        vector<vector<int>> ans;
        vector<int> subset;
        vector<pair<int, int>> freq;
        unordered_map<int, int> count;
        for (auto v : nums)
            count[v]++;
        // we will use freq vector instead of original nums vector and follow same logic as subsets
        for (auto p : count)
            freq.push_back(p);

        subsetsWithDup_helper(freq, 0, subset, ans);
        return ans;
    }
    void subsetsWithDup_helper(vector<pair<int, int>> &freq, int pos, vector<int> &subset, vector<vector<int>> &ans)
    {
        // we have processed all the elements
        if (pos == freq.size())
        {
            ans.push_back(subset);
            return;
        }

        // do not take it, all the cases where the current is not in subset
        subsetsWithDup_helper(freq, pos + 1, subset, ans);

        // take current if atleast 1 item is still present
        if (freq[pos].second > 0)
        {
            subset.push_back(freq[pos].first);
            // so it can be used at max remaining number of times
            freq[pos].second--;
            subsetsWithDup_helper(freq, pos, subset, ans);
            freq[pos].second++;
            subset.pop_back();
        }
    }
};

// 3
class CombinationSum
{
public:
    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {
        vector<vector<int>> all_combinations = {};
        vector<int> combination = {};
        combinationSum_helper(candidates, 0, target, combination, all_combinations);
        return all_combinations;
    }
    void combinationSum_helper(vector<int> &candidates, int pos, int target,
                               vector<int> &combination, vector<vector<int>> &all_combinations)
    {
        // we got one possible combination
        if (target == 0)
        {
            all_combinations.push_back(combination);
            return;
        }
        // we have processed all the elements and did not find a combination
        if (pos >= candidates.size())
            return;

        // do not take current
        combinationSum_helper(candidates, pos + 1, target, combination, all_combinations);

        // take current if it's value <= target
        if (candidates[pos] <= target)
        {
            combination.push_back(candidates[pos]);
            // pos will remain as it is, since we can take this item again
            combinationSum_helper(candidates, pos, target - candidates[pos], combination, all_combinations);
            combination.pop_back();
        }
    }
};

// 4
class CombinationSum2
{
public:
    vector<vector<int>> combinationSum2(vector<int> &candidates, int target)
    {
        vector<vector<int>> ans;
        vector<int> combination;
        unordered_map<int, int> count;
        vector<pair<int, int>> freq;
        for (auto v : candidates)
            count[v]++;
        // we will use freq vector instead of original nums vector and follow same logic as combination sum
        for (auto p : count)
            freq.push_back(p);
        combinationSum_helper(freq, 0, target, combination, ans);
        return ans;
    }
    void combinationSum_helper(vector<pair<int, int>> &freq, int pos, int target,
                               vector<int> &combination, vector<vector<int>> &ans)
    {
        // we got one possible combination
        if (target == 0)
        {
            ans.push_back(combination);
            return;
        }
        // we have processed all the elements and did not find a combination
        if (pos >= freq.size())
            return;

        // take current: if it's value <= target and it's freq is there
        if (freq[pos].first <= target && freq[pos].second > 0)
        {
            combination.push_back(freq[pos].first);
            freq[pos].second--;
            // keep the pos same, we might use same number again if it's freq is still >0
            combinationSum_helper(freq, pos, target - freq[pos].first, combination, ans);
            freq[pos].second++;
            combination.pop_back();
        }
        // do not take current
        combinationSum_helper(freq, pos + 1, target, combination, ans);
    }
};

// 7
class WordSearch
{
public:
    bool exist(vector<vector<char>> &board, string word)
    {
        for (int i = 0; i < board.size(); i++)
            for (int j = 0; j < board[0].size(); j++)
                if (search(board, i, j, word, 0))
                    return true;
        return false;
    }
    bool search(vector<vector<char>> &board, int i, int j, string word, int pos)
    {
        // all the letters have been matched already
        if (pos == word.size())
            return true;

        // out of range, alrady visited or word[index] is different
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] == '.' || board[i][j] != word[pos])
            return false;

        auto last = board[i][j];
        board[i][j] = '.'; // mark i,j as visited

        bool found = false;
        int dx[] = {0, 0, -1, 1};
        int dy[] = {1, -1, 0, 0};
        for (int k = 0; k < 4; k++)
            if (search(board, i + dx[k], j + dy[k], word, pos + 1))
            {
                found = true; // can' return true here since we need to correct the board before returning
                break;
            }

        board[i][j] = last; // reset the board
        return found;
    }
};
//

File: 9_Backtracking.py
from typing import List
from collections import Counter


# 1. https://leetcode.com/problems/subsets/
def subsets(self, nums: List[int]) -> List[List[int]]:
    # T:O(2^n) and S:O(n) T:O(n*2^n) if we consider the time to copy the subset to all_subsets as well

    def find_subset(index):
        if index == len(nums):
            all_subsets.append(cur_set[:])
            return

        # ignore current index
        find_subset(index + 1)
        # chooses the current index
        cur_set.append(nums[index])  # use it in cur subset
        find_subset(index + 1)
        cur_set.pop() # remove it from cur subset to as all combinations with num[index] in subset are already explored in above call

    all_subsets = []
    cur_set = []
    find_subset(0)
    return all_subsets

# 2. https://leetcode.com/problems/subsets-ii/
def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
    # T:O(2^n) and S:O(n) => T:(n*2^n) if we consider the time to copy the subset to all_subsets as well
    
    def find_subset(index):
        if index == len(freq_list):
            all_subsets.append(cur_subset[:])
        else:
            # ignore current index
            find_subset(index + 1)
            # ** use current index, if the number of items are still left and remain at same index
            if freq_list[index][1] > 0:
                freq_list[index][1] -= 1
                cur_subset.append(freq_list[index][0])
                find_subset(index)
                freq_list[index][1] += 1
                cur_subset.pop()

    # freq list, each element is [num, freq] and we will use the freq to decide how many times we can use the num in subset
    freq_list = [[k, v] for k, v in Counter(nums).items()]
    cur_subset = []
    all_subsets = []
    find_subset(0)
    return all_subsets


# 3. https://leetcode.com/problems/combination-sum/
def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
    # T:O(2^(n)) and S:O(n)
    def dfs(idx, target):
        if target == 0:
            all_comb.append(cur_comb[:])
            return
        if idx == len(candidates):
            return

        # **numbers are sorted all nums at >=idx are greater than target so can't be used
        if candidates[idx] > target:
            return

        # skip cur idx
        dfs(idx + 1, target)
        # choose cur idx and remain at same index to choose again
        cur_comb.append(candidates[idx])
        dfs(idx, target - candidates[idx]) 
        cur_comb.pop()

    candidates.sort()
    cur_comb = []
    all_comb = []
    dfs(0, target)
    return all_comb

# 4. https://leetcode.com/problems/combination-sum-ii/
def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
    # T:O(2^(n)) and S:O(n)
    def dfs(idx,target):
        if target==0:
            all_comb.append(cur_comb[:])
            return
        if idx==len(freq):
            return 
        
        # ** numbers are sorted all nums at >=idx are greater than target so can't be used
        if freq[idx][0] > target: 
            return 

        # skip cur idx
        dfs(idx+1,target)
        # choose cur idx if there are still some left, and remain at same index to choose again
        if freq[idx][1]>0:
            cur_comb.append(freq[idx][0])
            freq[idx][1]-=1
            dfs(idx,target-freq[idx][0])
            freq[idx][1]+=1
            cur_comb.pop()
    
    freq=[[k,v] for k,v in Counter(candidates).items()]
    freq.sort()
    cur_comb=[]
    all_comb=[]
    dfs(0,target)
    return all_comb


# 5. https://leetcode.com/problems/permutations/
def permute(self, nums: List[int]) -> List[List[int]]:
    # T:O(n*n!) and S:O(n) n! permutations and n for each perm to copy to all_perm
    def dfs(idx): #set num at pos idx
        if idx==n:
            all_perms.append(nums[:])
            return
        
        # for all nums idx to n, try all of them
        for i in range(idx,n):
            nums[idx],nums[i]=nums[i],nums[idx] #swap nums
            dfs(idx+1) #idx is set go for next index
            nums[idx],nums[i]=nums[i],nums[idx] #swap back to same state

    n=len(nums)
    all_perms=[]
    dfs(0)
    return all_perms

# 6. https://leetcode.com/problems/permutations-ii/
def permuteUnique(self, nums: List[int]) -> List[List[int]]:
    # T:O(n*n!) and S:O(n) n! permutations and n for each perm to copy to all_perm    
    def dfs(idx): #set num at pos idx
        if idx==n:
            all_perms.append(nums[:])
            return
        
        # for all nums idx to n, try all of them
        seen=set()
        for i in range(idx,n):
            if nums[i] not in seen: #don't try all permutations by setting same no twice at idx
                seen.add(nums[i])
                nums[idx],nums[i]=nums[i],nums[idx] #swap nums
                dfs(idx+1) #idx is set go for next index
                nums[idx],nums[i]=nums[i],nums[idx] #swap back to same state

    n=len(nums)
    all_perms=[]
    dfs(0)
    return all_perms


# 7. https://leetcode.com/problems/word-search/description/
def exist(self, board: List[List[str]], word: str) -> bool:
    # T: O(n*m*(4^w) and S:O(w) where w is length of word
    def dfs(i,j,idx):
        if idx==len(word): # Found the word
            return True
        if not (0<=i<n and 0<=j<m) or (i,j) in visited or word[idx]!=board[i][j]: # skip all invalid cases
            return False
        
        visited.add((i,j)) # mark the board[i][j] as visited
        for dx,dy in directions: # Try all 4 directions 
            if dfs(i+dx,j+dy,idx+1):
                return True
        visited.discard((i,j)) # unmark the board[i][j] as visited to use it in other combinations as well

    visited=set()
    n,m=len(board),len(board[0])
    directions=[[1,0],[-1,0],[0,1],[0,-1]]
    for i in range(n):
        for j in range(m):
            if dfs(i,j,0):
                return True
    return False

# 8. https://leetcode.com/problems/palindrome-partitioning/
def partition(self, s: str) -> List[List[str]]:
    # T:O(n*2^n) and S:O(n)
    # Time: O(n^2 * 2^(n-1)) ≈ O(n^2 * 2^n) — there are n-1 possible cut positions giving 2^(n-1) partition choices,
    # and for each state we may try O(n) substrings with each palindrome check costing O(n).
    def check_pal(i, j):
        while i <= j:
            if s[i] != s[j]:
                return False
            i += 1
            j -= 1
        return True

    def find_partition(index):
        if index == len(s): # reached end => one valid partition found
            all_parts.append(cur_parts[:])
            return

        # try every possible substring starting at index
        for i in range(index, len(s)):
            if check_pal(index, i): 
                cur_parts.append(s[index : i + 1]) # add the palindrome from index to i to current partition
                find_partition(i + 1) # look from i+1 for remaining partition
                cur_parts.pop()

    all_parts = []
    cur_parts = []
    find_partition(0)
    return all_parts


# 9. https://leetcode.com/problems/letter-combinations-of-a-phone-number/
def letterCombinations(self, digits: str) -> List[str]:
    # T:O(4^n) and S:O(n)
    # We have 4 choices for each digit and n digits so 4^n combinations, and n for each combination to copy to all_combs as well, so T:O(n*4^n)
    def dfs(idx):
        if idx==len(digits):
            all_combs.append("".join(cur))
            return
        
        for c in mapping[int(digits[idx])]:
            cur.append(c)
            dfs(idx+1)
            cur.pop()

    mapping=["","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]
    all_combs=[]
    cur=[]
    dfs(0)
    return all_combs
    
# 10. https://leetcode.com/problems/n-queens/
def solveNQueens(self, n: int) -> List[List[str]]:
    # Time: O(N!) — each row chooses among remaining valid columns, leading to at most N! placements.
    # Space: O(N) for recursion stack and column/diagonal tracking.
    def dfs(row):
        if row==n:
            new_board=["".join(board[i]) for i in range(n)]
            all_comb.append(new_board)
            return
        
        for col in range(n):
            # check we can put queen in [row,col] location
            if col_check[col] and left_dig[row+col] and right_dig[row-col+n-1]:
                col_check[col] = left_dig[row+col] = right_dig[row-col+n-1] = 0 # mark the column and diagonals as not available for next rows
                board[row][col]='Q' # place the queen and go for next row
                dfs(row+1)
                board[row][col]='.' # unplace the queen and mark the column and diagonals as available
                col_check[col] = left_dig[row+col] = right_dig[row-col+n-1] = 1

    col_check=[1]*n
    left_dig=[1]*(2*n-1)
    right_dig=[1]*(2*n-1)
    all_comb=[]
    board=[['.']*n for _ in range(n)]
    dfs(0)
    return all_comb

# 11. https://leetcode.com/problems/generate-parentheses/description/
def generateParenthesis(self, n: int) -> List[str]:
    """
    Correct: T:O(4^n/sqrt(n)) and S:O(n) auxiliary (recursion + current path), excluding output.
    
    from basic recursion: 2^2n = 4^n as we have 2n positions and at each position we have 2 choices
    
    Time:   O(Cn * n), where Cn is the nth Catalan number (number of valid parentheses strings).
            Since Cn ≈ 4^n / (n^(3/2)), time can also be written as O(4^n / sqrt(n)).
    """

    def dfs(left, right):
        if left == right == n: #computed one valid Parentheses
            all_combs.append("".join(cur))
            return

        # if left still remaining, choose it and generate from it
        if left < n:
            cur.append("(")
            dfs(left + 1, right)
            cur.pop()

        # if left>right, choose right as it is valid and generate from it
        if left > right:
            cur.append(")")
            dfs(left, right + 1)
            cur.pop()

    cur = []
    all_combs = []
    dfs(0, 0)
    return all_combs


# ################## Hard Problems ###################


# 1.  https://leetcode.com/problems/permutation-sequence/description/
def getPermutation(n: int, k: int) -> str:
    fact = [1] * n
    for i in range(1, n):
        fact[i] = fact[i - 1] * i

    k -= 1  # make it 0th indexed
    kth_perm = ""
    nums = [str(i) for i in range(1, n + 1)]

    while n:
        n -= 1  # compute this factorial
        index_to_use = k // fact[n]
        k = k % fact[n]
        kth_perm += nums[index_to_use]
        nums.pop(index_to_use)  # this number is already used in perm

    return kth_perm
